### DISCUSSION TITLE
Upgrade NPM and Node.JS Questions

### DISCUSSION THREAD

**Discussion by blaine07** [Ylianst/MeshCentral#5582] [Issues] [answered]
Afternoon,

I have had my MeshCentral server running on Ubuntu 20 for quite some time now. I apparently wasn't keeping it updated. I have node v10.19.0 and npm v6.14.4. I have been toying around with how to upgrade both, correctly, and I guess I can't seem to figure out why after upgrading I am left with MC not working.

Any suggested guides or advice on what I might be doing wrong updating NPM and node.js?

Today was trying this guide; for example: https://www.knowledgehut.com/blog/web-development/install-nodejs-on-ubuntu#how-to%C2%A0use-the%C2%A0node.js%C2%A0on-ubuntu%C2%A0

**Comment by si458** [comment, marked as answer]
it depends how you installed nodejs in the first place?
did you install it from the ubuntu repo `npm install nodejs` or did you install the nodesource then `npm install nodejs`?
personally, SNAPSHOT VM first then
1. i would stop meshcentral,
2. update nodejs to AT LEAST 18 using nodesource https://github.com/nodesource/distributions#installation-instructions
3. make sure the meshcentral service points to the new node (should be done automatically but no harm in checking) /etc/systemd/system/meshcentrl.service
4. start meshcentral again

npm will auto update to whatever version was included with node (unless u `npm install -g npm` to fully update npm but this isnt always needed

**Reply by blaine07** [reply]
Yeah, unfortunately, I have no idea which way/don't recall how I set node up initially.

Looking at your link it mentions 

NODE_MAJOR=20
echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR.x nodistro main" | sudo tee /etc/apt/sources.list.d/nodesource.list

How do I correctly fill out .."NODE_MAJOR" part for 18? Do I add 18 so it's ..."NODE_MAJOR=18" or do I replace "NODE_MAJOR" entirely?

**Reply by blaine07** [reply]
Okay, I have NPM and node upgraded. Figured out "NODE_MAJOR" thing. Once I restart MeshCentral it never comes back up though. How can I figure check where Node/npm now lie and correct it with MeshCentral?

**Reply by blaine07** [reply]
![image](https://github.com/Ylianst/MeshCentral/assets/63940698/cfe40548-2ae2-49a5-9679-e92c376a835d)

I think I have found it; but no idea how to get what this needs to be changed to to function correctly.

**Reply by si458** [reply]
check your `/etc/systemd/system/meshcentral.service` file to which node version it points too
my example `ExecStart=/usr/bin/node /home/simon/meshcentral/node_modules/meshcentral`
so you can then run `/usr/bin/node -v` to verify its using the latest node version

them check if the is any errors inside `meshcentral-data/mesherrors.txt`
also check `systemctl status meshcentral -n500`

**Reply by blaine07** [reply]
How do I find where new version of node is installed though? Yeah, see above, I found the meshcentral.service - just no idea what to change where node is now to?

**Reply by si458** [reply]
also if you run `node -v` and it returns (18/20), you can run `which node` and this will tell you where the node exec is located

**Reply by si458** [reply]
note i wouldnt put `node /home/blaine07...` i would use its full path `/usr/bin/node /home/blaine07...` if that makes sense?

**Reply by blaine07** [reply]
![image](https://github.com/Ylianst/MeshCentral/assets/63940698/bc2c416f-7154-4c90-a33b-ae331c0e23f5)

so, do I need to change anything in the meshcentral.service file at all?

![image](https://github.com/Ylianst/MeshCentral/assets/63940698/0d607fe2-7128-4561-8700-fb97f0d97211)

**Reply by si458** [reply]
erm no it should just work then?
did you stop meshcentral before u updated node?
is the a file inside `meshcentral-data` saying `mesherrors.txt`? whats the content of the file if the is?

**Reply by blaine07** [reply]
...what if I didn't stop meshcentral service before hand..sigh. Just now stopped and started service; looking for log now; one sec.

**Reply by blaine07** [reply]
![image](https://github.com/Ylianst/MeshCentral/assets/63940698/78517422-dc6d-4ce0-9edc-9cb3fee59b2d)
This tell you anything useful?

**Reply by blaine07** [reply]
that 930 would've been about when I got node upgraded I would suspect.

**Reply by si458** [reply]
mongodb issue, also you are running an old version, so i think u need to update to the new version,
1. stop meshcentral first!!!
2. go into `/home/blaine07` and then `npm install meshcentral` to update to a newer version
3. start meshcentral
4. check error logs again as if you get new errors should show 1.1.16

**Reply by blaine07** [reply]
No new errors in log but still not working.

I guess I am going to roll back my VM snapshop and start over from scratch, stopping the meshcentral service though..?

**Reply by si458** [reply]
is it the same logs you are seeing but showing 1.1.16 instead of 1.1.11?

but yes rollback if you can, then you could try updating ur meshcentral FIRST to latest then stopping meshcentral, updating nodejs, starting meshcentral?

**Reply by blaine07** [reply]
No nothing new in log at all; just same as above for the 0930 time stamp.

Yes, rolling back now. I tried updating MeshCentral first the other day but it wouldn't work because NODE was too old. Will do above though/ update meshcentral, stop meshcentral, update nodejs and starting

**Reply by si458** [reply]
oh hang fire! it might of been installing new modules! which can take upto 3 minutes to start! then other restarts afterwards would be quicker
you could of checked using `systemctl status meshcentral` and see what the last few lines where
DOH sorry ! having a brain fart of a sunday!

**Reply by blaine07** [reply]
NO BIGGIE! Already almost restarted snapshop(Xcp-NG).. Will report back..shortly.

**Reply by blaine07** [reply]
![image](https://github.com/Ylianst/MeshCentral/assets/63940698/5501e36e-2213-4cec-8ed8-0beaa21cb269)
(more lines here but redacted because my domain)
maybe still loading; giving a bit longer?

![image](https://github.com/Ylianst/MeshCentral/assets/63940698/d4e75fec-dca0-4b61-a2cc-ab279365c130)


So far nothing though; not working.

**Reply by blaine07** [reply]
WINNER WINNER CHICKEN DINNER - its loaded and running :wipes brow:

Thank you SO SO SO MUCH!

Do you happen to know anything personally about this black screen/cloudflare thing? Everything I have runs through cloudflare and not much way around it so stinks it's causing this "black screen" thing?

cloudflare issues: multiple attempts/black screen · Issue #5302 · Ylianst/MeshCentral · GitHub

https://github.com/Ylianst/MeshCentral/issues/5302

**Reply by si458** [reply]
sadly not, ive tried debugging it over and over, and it always hitting the same issue now
where its now connecting ok, but when it starts to send the KVM(video) pictures to cloudflare, cloudflare just ignores it?
no idea why?
i would keep following #5302 for the moment
and the work around for the moment is to TURN OFF the dns proxy setting and let your clients/remote devices talk directly to your server,
but this wont work if you are behind a firewall and not comfy opening ports up and forwarding etc (like cloudflare tunnels)
